OneCommunityGlobal/HGNRest
# Abi weekly summary reports page

# Description
This PR implements a new aggregation pipeline for volunteer trends data as well as fixes previous queries for the volunteer hours stats and work distribution stats

New URL format: *`api/reports/volunteerstats?startDate=2024-05-26&endDate=2024-06-02&comparisonStartDate=2024-05-18&comparisonEndDate=2024-05-25`*

## Changes 9/13
* Updated API to retrieve comparison data for the following queries:
   * `totalActiveTeams`
   * `totalBadgesAwarded`
   * `totalHoursWorked`

```javascript
   "totalBadgesAwarded": {
        "current": 40,
        "comparison": 40,
        "percentage": 0
    },
    "totalActiveTeams": {
        "current": 339,
        "comparison": 330,
        "percentage": 0.03
    },
   "totalHoursWorked": {
        "current": 161.93333333333334,
        "comparison": 342.56666666666666,
        "percentage": -0.53
    },
```

## Changes 8/23
* Updated the api to require a comparisonStartDate and comparisonEndDate
  * This comparison time range is the period in time that will be compared to the current week
* Created the comparison data for the Volunteer Numbers Stats
  * The new data will be formatted like so:
```javascript
"volunteerNumberStats": {
        "activeVolunteers": {
            "count": 1373,
            "comparisonPercentage": 0.01
        },
        "newVolunteers": {
            "count": 13,
            "comparisonPercentage": -0.32
        },
        "deactivatedVolunteers": {
            "count": 2,
            "comparisonPercentage": 1
        }
    },
```

**NOTE** the  for the `comparisonPercentage` number. -0.32 means a decrese of 32% from the comparison week. 1 is a 100% increase

## Related PRS (if any):
This PR is built off of #994 

## Main changes explained:
- Created a new query to aggregate the monthly volunteer trends data for the past year.
- Fixed bugs for the volunteer hours stats and work distribution stats to now return correct data
…

## How to test:
1. check into current branch
2. do `npm install` and `npm run dev`
3. Log in with an owner credential to Postman via the endpoint `POST http://localhost:4500/api/login` and paste the JWT token into the headers tab as shown below
4. Query the endpoint with Postman: `api/reports/volunteerstats?startDate=2024-05-26&endDate=2024-06-02&comparisonStartDate=2024-05-18&comparisonEndDate=2024-05-25`
5. verify the data is correct
![Screenshot 2024-08-12 at 5 00 53 PM](https://github.com/user-attachments/assets/19f7c29a-8bc0-4e64-b2ed-c8b47cb5a335)


## Screenshots or videos of changes:
![Screenshot 2024-08-12 at 4 55 30 PM](https://github.com/user-attachments/assets/883d6abd-0da5-4135-8a5d-4329769d1f22)
![Screenshot 2024-08-12 at 4 55 48 PM](https://github.com/user-attachments/assets/9fdbd73c-16f0-417c-a808-7e77ed1e58d3)
